File: BraintreeCore/src/main/java/com/braintreepayments/api/ApiClient.java
package com.braintreepayments.api;

import androidx.annotation.VisibleForTesting;

import org.json.JSONException;
import org.json.JSONObject;

import java.lang.ref.WeakReference;

class ApiClient {

    static final String PAYMENT_METHOD_ENDPOINT = "payment_methods";

    private final WeakReference<BraintreeClient> braintreeClientRef;

    ApiClient(BraintreeClient braintreeClient) {
        this(new WeakReference<>(braintreeClient));
    }

    @VisibleForTesting
    ApiClient(WeakReference<BraintreeClient> braintreeClientRef) {
        this.braintreeClientRef = braintreeClientRef;
    }

    void tokenizeGraphQL(final JSONObject tokenizePayload, final TokenizeCallback callback) {
        final BraintreeClient braintreeClient = braintreeClientRef.get();
        if (braintreeClient == null) {
            return;
        }

        braintreeClient.sendAnalyticsEvent("card.graphql.tokenization.started");
        braintreeClient.sendGraphQLPOST(tokenizePayload.toString(), new HttpResponseCallback() {

            @Override
            public void onResult(String responseBody, Exception httpError) {
                if (responseBody != null) {
                    try {
                        callback.onResult(new JSONObject(responseBody), null);
                        braintreeClient.sendAnalyticsEvent("card.graphql.tokenization.success");
                    } catch (JSONException exception) {
                        braintreeClient.sendAnalyticsEvent("card.graphql.tokenization.failure");
                        callback.onResult(null, exception);
                    }
                } else {
                    braintreeClient.sendAnalyticsEvent("card.graphql.tokenization.failure");
                    callback.onResult(null, httpError);
                }
            }
        });
    }

    void tokenizeREST(final PaymentMethod paymentMethod, final TokenizeCallback callback) {
        final BraintreeClient braintreeClient = braintreeClientRef.get();
        if (braintreeClient == null) {
            return;
        }

        String url = ApiClient.versionedPath(
                ApiClient.PAYMENT_METHOD_ENDPOINT + "/" + paymentMethod.getApiPath());

        paymentMethod.setSessionId(braintreeClient.getSessionId());

        try {
            braintreeClient.sendPOST(url, paymentMethod.buildJSON().toString(), new HttpResponseCallback() {

                @Override
                public void onResult(String responseBody, Exception httpError) {
                    if (responseBody != null) {
                        try {
                            callback.onResult(new JSONObject(responseBody), null);
                        } catch (JSONException exception) {
                            callback.onResult(null, exception);
                        }
                    } else {
                        callback.onResult(null, httpError);
                    }
                }
            });
        } catch (JSONException exception) {
            callback.onResult(null, exception);
        }
    }

    static String versionedPath(String path) {
        return "/v1/" + path;
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/ApiClient.kt
package com.braintreepayments.api

import androidx.annotation.RestrictTo
import org.json.JSONException
import org.json.JSONObject

/**
 * @suppress
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class ApiClient(private val braintreeClient: BraintreeClient) {

    fun tokenizeGraphQL(tokenizePayload: JSONObject, callback: TokenizeCallback) =
        braintreeClient.run {
            sendAnalyticsEvent("card.graphql.tokenization.started")
            sendGraphQLPOST(tokenizePayload.toString(), object : HttpResponseCallback {
                override fun onResult(responseBody: String?, httpError: Exception?) {
                    parseResponseToJSON(responseBody)?.let { json ->
                        sendAnalyticsEvent("card.graphql.tokenization.success")
                        callback.onResult(json, null)
                    } ?: httpError?.let { error ->
                        sendAnalyticsEvent("card.graphql.tokenization.failure")
                        callback.onResult(null, error)
                    }
                }
            })
        }

    fun tokenizeREST(paymentMethod: PaymentMethod, callback: TokenizeCallback) =
        braintreeClient.run {
            val url = versionedPath("$PAYMENT_METHOD_ENDPOINT/${paymentMethod.apiPath}")
            paymentMethod.setSessionId(braintreeClient.sessionId)

            sendAnalyticsEvent("card.rest.tokenization.started")
            sendPOST(url, paymentMethod.buildJSON().toString(), object : HttpResponseCallback {
                override fun onResult(responseBody: String?, httpError: Exception?) {
                    parseResponseToJSON(responseBody)?.let { json ->
                        sendAnalyticsEvent("card.rest.tokenization.success")
                        callback.onResult(json, null)
                    } ?: httpError?.let { error ->
                        sendAnalyticsEvent("card.rest.tokenization.failure")
                        callback.onResult(null, error)
                    }
                }
            })
        }

    private fun parseResponseToJSON(responseBody: String?): JSONObject? =
        responseBody?.let {
            try {
                JSONObject(it)
            } catch (e: JSONException) {
                null
            }
        }

    companion object {
        const val PAYMENT_METHOD_ENDPOINT = "payment_methods"

        @JvmStatic
        fun versionedPath(path: String): String {
            return "/v1/$path"
        }
    }
}
